src/main/java/com/zendesk/maxwell/schema/MysqlSavedSchema.java

```diff
@@ -761,4 +754,199 @@ public static String getSchemaPositionSHA(Long serverID, Position position) {
 		);
 		return DigestUtils.shaHex(shaString);
 	}
+
+	public int getSchemaChainLength() {
+		return schemaChain.length();
+	}
+
+	public int compact(Connection connection, int chainMaxLength, int chainLength) throws SQLException {
+		if (chainMaxLength < 0) {
+			return 0;
+		}
+
+		assert chainLength <= chainMaxLength;
+
+		if (schemaID == null) {
+			throw new RuntimeException("Schema not saved yet.");
+		}
+		if (schemaChain.length() <= chainLength || schemaChain.length() < chainMaxLength) {
+			return 0;
+		}
+
+		long[] removeIds;
+		SchemaChain newSchemaChain = new SchemaChain();
+
+		connection.setAutoCommit(false);
+		try {
+			Long newBaseSchemaID;
+
+			if (chainLength == 0) {
+				// remove all base schemas
+
+				removeIds = new long[schemaChain.length()];
+				int n = 0;
+				for (long id : schemaChain) {
+					removeIds[n++] = id;
+				}
+				newBaseSchemaID = schemaID;
+				saveFullSchemaData(connection, schemaID);
+
+			} else {
+				// split chain between schemas to keep and to remove
+
+				long[] keepIds = new long[chainLength];
+				removeIds = new long[schemaChain.length() - chainLength];
+				int n = 0;
+				for (long id : schemaChain) {
+					if (n < chainLength) {
+						keepIds[n++] = id;
+					} else {
+						removeIds[n++ - chainLength] = id;
+					}
+				}
+
+				// save new base schema data
+				newBaseSchemaID = keepIds[chainLength - 1];
+
+				MysqlSavedSchema newBase = new MysqlSavedSchema(serverID, sensitivity);
+				try {
+					newBase.restoreFromSchemaID(connection, newBaseSchemaID);
+				} catch (InvalidSchemaError e) {
+					throw new RuntimeException("Couldn't restore schema " + newBaseSchemaID + ".", e);
+				}
+
+				newBase.saveFullSchemaData(connection, newBase.schemaID);
+
+				for (int i = keepIds.length - 1; i >= 0; i--) {
+					newSchemaChain.push(keepIds[i]);
+				}
+			}
+
+			// remove reference to parent of the new base
+			try (PreparedStatement stmt = connection.prepareStatement(
+				"UPDATE `schemas` SET `base_schema_id` = NULL, `deltas` = NULL WHERE `id`= ?"
+			)) {
+				stmt.setLong(1, newBaseSchemaID);
+				stmt.executeUpdate();
+			}
+
+			// clean up
+			StringBuilder inSQL = new StringBuilder();
+			for (int i = 0; i < removeIds.length; i++) {
+				if (inSQL.length() > 0) {
+					inSQL.append(", ");
+				}
+				inSQL.append("?");
+			}
+			for (String tableName : new String[]{"columns", "tables", "databases"}) {
+				try (PreparedStatement stmt = connection.prepareStatement(
+					"DELETE FROM `" + tableName + "` WHERE `schema_id` IN (" + inSQL + ")"
+				)) {
+					for (int i = 0; i < removeIds.length; i++) {
+						stmt.setLong(i + 1, removeIds[i]);
+					}
+					stmt.executeUpdate();
+				}
+			}
+
+			try (PreparedStatement stmt = connection.prepareStatement(
+				"DELETE FROM `schemas` WHERE `id` IN (" + inSQL + ")"
+			)) {
+				for (int i = 0; i < removeIds.length; i++) {
+					stmt.setLong(i + 1, removeIds[i]);
+				}
+				stmt.executeUpdate();
+			}
+
+			connection.commit();
```

Comment:
this method is really really long, let's at least break it up into two sub-methods, compaction and cleanup.

Reply:
After removing code for removing only part of the chain, I believe this is not necessary anymore?

Reply:
no, I'd still like you to break it in two.